packages/cardano-services/test/Asset/TypeormAssetProvider/TypeormAssetProvider.test.ts

```diff
@@ -0,0 +1,126 @@
+import { logger } from '@cardano-sdk/util-dev';
+
+const tokenMetadata = {
+  assetId: '4abc',
+  decimals: 8,
+  desc: 'SingularityNET',
+  icon: 'testLogo',
+  name: 'SingularityNet AGIX Token',
+  ticker: 'AGIX',
+  url: 'https://singularitynet.io/'
+};
+
+describe('TypeormAssetProvider', () => {
+  let provider: TypeormAssetProvider;
+  let fixtureBuilder: TypeormAssetFixtureBuilder;
+
+  beforeEach(async () => {
+    const dnsResolver = createDnsResolver({ factor: 1.1, maxRetryTime: 1000 }, logger);
+    const entities = getEntities(['asset']);
+    const connectionConfig$ = getConnectionConfig(dnsResolver, 'test', 'Handle', {
+      postgresConnectionStringHandle: process.env.POSTGRES_CONNECTION_STRING_HANDLE!
+    });
+    const tokenMetadataService: TokenMetadataService = {
+      getTokenMetadata: jest.fn().mockResolvedValue([tokenMetadata]),
+      shutdown: jest.fn()
+    };
+
+    provider = new TypeormAssetProvider(
+      {
+        paginationPageSizeLimit: PAGINATION_PAGE_SIZE_LIMIT_ASSETS
+      },
+      { connectionConfig$, entities, logger, tokenMetadataService }
+    );
+    fixtureBuilder = new TypeormAssetFixtureBuilder({ connectionConfig$, entities, logger });
+
+    await provider.initialize();
+    await provider.start();
+    await fixtureBuilder.initialize();
+    await fixtureBuilder.start();
+  });
+
+  afterEach(async () => {
+    jest.restoreAllMocks();
+
+    await provider.shutdown();
+    await fixtureBuilder.shutdown();
+  });
+
+  describe('getAsset', () => {
+    it('should return asset info for a given asset id', async () => {
+      const assetQuery = await fixtureBuilder.getAssets(1);
+      const testAsset = assetQuery[0];
+      const asset = await provider.getAsset({ assetId: testAsset.assetId });
+      expect(asset).toEqual(testAsset);
+    });
+
+    it('should throw an error if the asset does not exist', async () => {
+      const assetId = Cardano.AssetId('64190c10b567ad4bcf254ad0f76eb374d749d0b25fd82786af6a839a48656c6c6f48616e646c65');
+      await expect(provider.getAsset({ assetId })).rejects.toThrowError(ProviderError);
+    });
+
+    it('should return metadata if the asset has metadata', async () => {
+      const assetQuery = await fixtureBuilder.getAssets(1, { with: [TypeormAssetWith.metadata] });
+      const testAsset = assetQuery[0];
+      const asset = await provider.getAsset({ assetId: testAsset.assetId, extraData: { nftMetadata: true } });
+      expect(asset).toEqual(testAsset);
+    });
+
+    it('should return undefined asset token metadata if the token registry throws a server internal error', async () => {
+      const assetQuery = await fixtureBuilder.getAssets(1, { with: [TypeormAssetWith.metadata] });
+      const testAsset = assetQuery[0];
+      const asset = await provider.getAsset({
+        assetId: testAsset.assetId,
+        extraData: { nftMetadata: true, tokenMetadata: true }
+      });
+      expect(asset.tokenMetadata).toEqual(tokenMetadata);
```

Comment:
This test seems to be missing relevant setup too
```suggestion
      expect(asset.tokenMetadata).toBeUndefined();
```

Reply:
This test was setup [here](https://github.com/input-output-hk/cardano-js-sdk/blob/7d7c8a101bc029b35e4bc332dc4b800aaae32563/packages/cardano-services/test/Asset/TypeormAssetProvider/TypeormAssetProvider.test.ts#L33-L36). So every `tokenMetadata` should have [this](https://github.com/input-output-hk/cardano-js-sdk/blob/7d7c8a101bc029b35e4bc332dc4b800aaae32563/packages/cardano-services/test/Asset/TypeormAssetProvider/TypeormAssetProvider.test.ts#L13-L21) data

Reply:
It's not testing what it says, and we need to test it: `should return undefined asset token metadata if the token registry throws a server internal error`

Reply:
Aha. Got you. Working on that.

Reply:
Fixed on fc7f8c4

Reply:
It would be cleaner to not re-create the `TypeormAssetProvider` in this test: use a `tokenMetadataService` variable in the outer scope (`describe('TypeormAssetProvider'`) and just re-configure mock behavior in this test.

Reply:
Do you mean something like this? 468d6f7

Reply:
Using `Jest.Mocked<TokenMetadataService>` here: 74de9de